core/services/pipeline/graph.go

```diff
@@ -37,9 +38,35 @@ func (g *Graph) UnmarshalText(bs []byte) (err error) {
 	if err != nil {
 		return errors.Wrap(err, "could not unmarshal DOT into a pipeline.Graph")
 	}
+	g.AddImplicitDependenciesAsEdges()
 	return nil
 }
 
+// Looks at node attributes and searches for implicit dependencies on other nodes
+// expressed as attribute values. Adds those dependencies as edges in the graph.
+func (g *Graph) AddImplicitDependenciesAsEdges() {
+	for nodesIter := g.Nodes(); nodesIter.Next(); {
+		graphNode := nodesIter.Node().(*GraphNode)
+
+		params := make(map[string]bool)
+		// Walk through all attributes and find all params which this node depends on
+		for _, attr := range graphNode.Attributes() {
+			for _, item := range variableRegexp.FindAll([]byte(attr.Value), -1) {
```

Comment:
I'm guessing that we're going to have to explicitly handle the "magical" variables injected by the job runner code -- `$(jobSpec.xxxx)` and `$(jobRun.xxxx)`.  For reference, you can find where each job type's code injects these magical variables by searching for `"jobRun"` or `"jobSpec"` in the subpackages of `core/services/`.

Reply:
The magical variables don't match any task names, so are just ignored. I think we don't need any other handling for those.